system/known_hosts.py

```diff
@@ -0,0 +1,234 @@
+    #trailing newline in files gets lost, so re-add if necessary
+    if key is not None and key[-1]!='\n':
+        key+='\n'
+
+    if key is None and state != "absent":
+        module.fail_json(msg="No key specified when adding a host")
+
+    sanity_check(module,host,key)
+
+    current,replace=search_for_host_key(module,host,key,path)
+
+    #We will change state if current==True & state!="present"
+    #or current==False & state=="present"
+    #i.e (current) XOR (state=="present")
+    #Alternatively, if replace is true (i.e. key present, and we must change it)
+    if module.check_mode:
+        module.exit_json(changed = replace or ((state=="present") != current))
+
+    #Now do the work.
+
+    #First, remove an extant entry if required
+    if replace==True or (current==True and state=="absent"):
+        module.run_command(['ssh-keygen','-R',host,'-f',path],
+                           check_rc=True)
+        params['changed'] = True
+    #Next, add a new (or replacing) entry
+    if replace==True or (current==False and state=="present"):
+        try:
+            inf=open(path,"r")
+        except IOError, e:
+            if e.errno == errno.ENOENT:
+                inf=None
+            else:
+                module.fail_json(msg="Failed to read %s: %s" % \
+                                     (path,str(e)))
+        try:
+            outf=tempfile.NamedTemporaryFile(dir=os.path.dirname(path),
+                                             delete=False)
+            if inf is not None:
+                for line in inf:
+                    outf.write(line)
+                inf.close()
+            outf.write(key)
+            outf.close()
+            module.atomic_move(outf.name,path)
+        except IOError,e:
+            module.fail_json(msg="Failed to write to file %s: %s" % \
+                                 (path,str(e)))
+        params['changed'] = True
+    
+    return params
+
+def sanity_check(module,host,key):
+    '''Check supplied key is sensible
+
+    host and key are parameters provided by the user; If the host
+    provided is inconsistent with the key supplied, then this function
+    quits, providing an error to the user.
+    '''
+    #If no key supplied, we're doing a removal, and have nothing to check here.
+    if key is None:
+        return
+    #Rather than parsing the key ourselves, get ssh-keygen to do it
+    #(this is essential for hashed keys, but otherwise useful, as the
+    #key question is whether ssh-keygen thinks the key matches the host).
+
+    #The approach is to write the key to a temporary file,
+    #and then attempt to look up the specified host in that file.
+    try:
+        outf=tempfile.NamedTemporaryFile(delete=False)
+        outf.write(key)
+        outf.close()
+    except IOError,e:
+        module.fail_json(msg="Failed to write to temporary file %s: %s" % \
+                             (outf.name,str(e)))
+    rc,stdout,stderr=module.run_command(['ssh-keygen','-F',host,
```

Comment:
Thanks for these comments, which I'll get to shortly. Would you like a commit on top of this commit (for a 2-commit PR) or for me to incorporate changes and still leave a single commit here?

To answer your second question, because ssh-keyscan is in general insecure (it doesn't do any host validation).